sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1698,23 +1729,35 @@ static void mc_dailink_exit_loop(struct snd_soc_card *card)
 {
 	struct snd_soc_dai_link *link;
 	int ret;
-	int i, j;
+	int i, j, k, m;
+	bool found;
 
 	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++) {
-		if (!codec_info_list[i].dais[0].exit)
-			continue;
-		/*
-		 * We don't need to call .exit function if there is no matched
-		 * dai link found.
-		 */
-		for_each_card_prelinks(card, j, link) {
-			if (!strcmp(link->codecs[0].dai_name,
-				    codec_info_list[i].dais[0].dai_name)) {
-				ret = codec_info_list[i].dais[0].exit(card, link);
-				if (ret)
-					dev_warn(card->dev,
-						 "codec exit failed %d\n",
-						 ret);
+		found = false;
+		for (k = 0; k < codec_info_list[i].dai_num; k++) {
+			if (!codec_info_list[i].dais[k].exit)
+				continue;
+			/*
+			 * We don't need to call .exit function if there is no matched
+			 * dai link found.
+			 */
+			for_each_card_prelinks(card, j, link) {
+				for (m = 0; m < link->num_codecs; m++) {
+					if (!strcmp(link->codecs[m].dai_name,
+						   codec_info_list[i].dais[k].dai_name)) {
+						found = true;
+						goto run_exit;
+					}
+				}
+			}
+		}
+run_exit:
+		if (found) {
+			ret = codec_info_list[i].dais[k].exit(card, link);
+			if (ret) {
+				dev_warn(card->dev,
+					 "codec exit failed %d\n",
+					 ret);
```

Comment:
I cant say that I understood exactly what's going on with the 4-level nested loop but it doesnt look pretty. I dont have an idea of how to make it better though

Reply:
> I cant say that I understood exactly what's going on with the 4-level nested loop but it doesnt look pretty. I dont have an idea of how to make it better though

I added some comments in the code. Hope it helps.

Reply:
@ranj063 Thanks for the suggestion. I did slight change because we need to check codec_info_list one by one instead of checking each link. Because a codec dai could be use twice in the link, and we will call .exit function twice if we call .exit for each link.

Reply:
but thats why I had the break inside the if (exit). I think logically it makes to have the for_each_card_prelinks(card, i, link) as the outermost loop but I am not going to block the PR because of that

Reply:
> but thats why I had the break inside the if (exit). I think logically it makes to have the for_each_card_prelinks(card, i, link) as the outermost loop but I am not going to block the PR because of that

The codec could be in different link, and we can have different codecs in the same link.
For example, 
SDW0-Plyaback uses rt711 for headset playback
SDW0-Captue uses rt711 for headset capture
SDW2-Playback uses rt1316 and rt1308 for speaker playback.
With the break inside the if(exit), rt711 .exit will be called twice, and rt1308 .exit will not be called.